services/src/main/java/org/keycloak/broker/oidc/AbstractOAuth2IdentityProvider.java

```diff
@@ -327,6 +332,15 @@ protected UriBuilder createAuthorizationUrl(AuthenticationRequest request) {
         if (acr != null) {
             uriBuilder.queryParam(OAuth2Constants.ACR_VALUES, acr);
         }
+        String forwardParameterConfig = getConfig().getForwardParameters() != null ? getConfig().getForwardParameters(): "";
+        List<String> forwardParameters = Arrays.asList(forwardParameterConfig.split("\\s*,\\s*"));
+        for(String forwardParameter: forwardParameters) {
+            String name = AuthorizationEndpoint.LOGIN_SESSION_NOTE_ADDITIONAL_REQ_PARAMS_PREFIX + forwardParameter.trim();
```

Comment:
I think this will support just forwarding of custom parameters not defined in the specs (EG. parameter "forwarded_parameter" you used in the test), but no forwarding of standard parameters (EG. "nonce"). If we want to support also forwarding of standard parameters (which IMO we should), this line can be potentially refactored to something like:
```
String forwardParameter = forwardParameter.trim();
String name = AuthzEndpointRequestParser.KNOWN_REQ_PARAMS.contains(forwardParameter) ? forwardParameter : AuthorizationEndpoint.LOGIN_SESSION_NOTE_ADDITIONAL_REQ_PARAMS_PREFIX + forwardParameter;
```

Also the test would be good for some of standard parameters (EG. nonce).

Reply:
In my opinion, standard parameters should not be **simply** forwarded to avoid errors in external IdP.

- Duplicate parameters

For example, if "client_id" is specified in configuration, keycloak send an own "client_id" and a client requested "client_id". I think most of external IdP cannot handle these request.

- Default value

There is some difference in handling default values. ignoring parameters, generating parameters or using parameters from the configuration ("nonce" is generated if not specified ). I think we need to handle default behavior for each parameter.

I plan to create another PR to forward standard parameters.

Reply:
Yes, for "client_id" it probably doesn't have much sense, but for some others, it maybe has? Also the UriBuilder has method "replaceQueryParam" to avoid duplicated query parameters. But on other hand, in AbstractOAuth2IdentityProvider.createAuthorizationUrl method, I see that most of the parameters where forwarding makes sense (prompt, nonce, acr_values) are already automatically forwarded to IDP. So maybe just improve tooltip that forwarding is supported just for non-standard OpenID Connect parameters and leave support just for custom parameters?

Reply:
I will improve the tooltip message as follows.

`Non OpenID Connect/OAuth standard query parameters to be forwarded to external IDP from the initial application request to Authorization Endpoint. Multiple parameters can be entered, separated by comma (,).`

Does this make sense?

Reply:
@tmogi001 +1